WordPress/src/test/java/org/wordpress/android/ui/uploads/LocalDraftUploadStarterTest.kt

```diff
@@ -0,0 +1,223 @@
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.mockito.junit.MockitoJUnitRunner
+import org.wordpress.android.fluxc.model.PostModel
+import org.wordpress.android.fluxc.model.SiteModel
+import org.wordpress.android.fluxc.store.PostStore
+import org.wordpress.android.fluxc.store.SiteStore
+import org.wordpress.android.util.NetworkUtilsWrapper
+import org.wordpress.android.viewmodel.helpers.ConnectionStatus
+import org.wordpress.android.viewmodel.helpers.ConnectionStatus.AVAILABLE
+import org.wordpress.android.viewmodel.helpers.ConnectionStatus.UNAVAILABLE
+import java.util.UUID
+import kotlin.random.Random
+
+@RunWith(MockitoJUnitRunner::class)
+class LocalDraftUploadStarterTest {
+    @get:Rule val rule = InstantTaskExecutorRule()
+
+    private val sites = listOf(SiteModel(), SiteModel())
+    private val sitesAndPosts: Map<SiteModel, List<PostModel>> = mapOf(
+            sites[0] to listOf(createPostModel(), createPostModel()),
+            sites[1] to listOf(
+                    createPostModel(),
+                    createPostModel(),
+                    createPostModel(),
+                    createPostModel(),
+                    createPostModel()
+            )
+    )
+    private val posts = sitesAndPosts.values.flatten()
+
+    private val siteStore = mock<SiteStore> {
+        on { sites } doReturn sites
+    }
+    private val postStore = mock<PostStore> {
+        sites.forEach {
+            on { getLocalDraftPosts(eq(it)) } doReturn sitesAndPosts[it]
+        }
+    }
+
+    @Test
+    fun `when the internet connection is restored, it uploads all local drafts`() {
+        // Given
+        val connectionStatus = createConnectionStatusLiveData(UNAVAILABLE)
+        val uploadServiceFacade = createMockedUploadServiceFacade()
+
+        val starter = createLocalDraftUploadStarter(connectionStatus, uploadServiceFacade)
+        starter.activateAutoUploading(createMockedProcessLifecycleOwner())
+
+        // When
+        runBlocking {
+            connectionStatus.postValue(AVAILABLE)
+
+            starter.waitForAllCoroutinesToFinish()
+        }
+
+        // Then
+        verify(uploadServiceFacade, times(posts.size)).uploadPost(
+                context = any(),
+                post = any(),
+                trackAnalytics = any(),
+                publish = any(),
+                isRetry = eq(true)
+        )
+    }
+
+    @Test
+    fun `when the app is placed in the foreground, it uploads all local drafts`() {
+        // Given
+        val connectionStatus = createConnectionStatusLiveData(AVAILABLE)
+        val uploadServiceFacade = createMockedUploadServiceFacade()
+
+        val lifecycle = LifecycleRegistry(mock()).apply { handleLifecycleEvent(Event.ON_CREATE) }
+
+        val starter = createLocalDraftUploadStarter(connectionStatus, uploadServiceFacade)
+        starter.activateAutoUploading(createMockedProcessLifecycleOwner(lifecycle))
+
+        // When
+        runBlocking {
```

Comment:
I'm not sure. The reason I did it this way is to cover the structured concurrency as well. If everything is sequential, then I can't prove that the coroutine usage was correct and really concurrent. I could have made a mistake and I have nothing to protect me from that. 

For example, if you remove the `scope` from `scope.launch` in `LocalDraftUploadStarter`:

```kt
    private fun upload(scope: CoroutineScope, site: SiteModel) = scope.launch(ioDispatcher) {

    // to this:
    private fun upload(scope: CoroutineScope, site: SiteModel) = launch(ioDispatcher) {
```

These tests would fail. After applying your suggestions, they pass. But the coroutine usage is incorrect. 🙂 

> At the same time I believe having deterministic unit tests is really important

Doesn't the `runBlocking` and `waitForAllCoroutinesToFinish()` make this deterministic? I haven't used the term "deterministic" much so let me know if I'm wrong. 😄

Reply:
>Doesn't the runBlocking and waitForAllCoroutinesToFinish() make this deterministic? I haven't used the term "deterministic" much so let me know if I'm wrong. 😄

I'll use a definition from Wikipedia `a deterministic system is a system in which no randomness is involved in the development of future states of the system.`. I think `runBlocking` makes sure the current thread doesn't continue until the code in the block is completed. However, if I understand it correctly the order in which `upload` coroutines are execute is non-deterministic (in other words it's random and if we run the test 100 times it won't always be the same -> it's prone to flakiness). 
Imagine for example you are refactoring the code and you run the test after each change -> if the test fails you assume the change broke the test, however if the test is non-deterministic you can never be sure as a week old change might have caused it.
I'd probably suggest having two separate tests. One deterministic test in this class and a non-deterministic version with explanation in a separate class. 
@planarvoid You have more experience with both coroutines and unit-tests. Wdyt?

Reply:
I don't have the full context so just consider this a suggestion from a quick glance at the code. In my opinion a unit test should run the same way every time it's triggered with consistent results. I don't think unit tests should run in a concurrent environment because we can't ensure consistent results. Checking your code I see that you access directly `coroutineContext` from the test and it's public only because of the test (which might be a bad sign). I think injecting the `Unconfined` context in the test is the right way to go. This way you can still check all the jobs will finish correctly. You're just not checking the parallel execution. Testing parallel execution is not easy anyway. All of the jobs could be running on a single thread and you wouldn't notice in your test (and I think that's fine). 
One more thing I'd consider is moving the code that triggers the parallel upload (inside the `checkConnectionAndUpload` method) into a separate class. This way you could really unit test that piece of logic. 
And a last thing, what do you think about moving the `corouroutineScope` from a parameter of the `upload` method and simply calling `launch { upload(site) }`. I find it much more readable than passing context as a parameter.

Reply:
@planarvoid Thank you for responding!

> And a last thing, what do you think about moving the corouroutineScope from a parameter of the upload method and simply calling launch { upload(site) }. I find it much more readable than passing context as a parameter.

Please see [this discussion](https://github.com/wordpress-mobile/WordPress-Android/pull/9869#discussion_r285019477) for why it's like that. 

> Checking your code I see that you access directly coroutineContext from the test and it's public only because of the test (which might be a bad sign)

The `coroutineContext` is public because I opted to make `LocalDraftUploadStarter` implement `CoroutineScope` instead of relying on a global scope.

Reply:
@planarvoid 

> Please see this discussion for why it's like that.

I'm sorry. I responded without analyzing what you said. You're right. That may be better. I'll look at it.

Reply:
@malinajirka I've applied what's been discussed. 

- [x] 2c6c2f28dadf0be09e70ead16c8de523912408a9 Removed concurrency from `LocalDraftUploadStarterTest`
- [x] 3a11c0ebf45569fcaf31f2df2db8cae9f909d670 Added `LocalDraftUploadStarterConcurrentTest`
- [x] 3f384f1d76998a80d8ec9dcaa61677a6b2cb9393 Simplified `LocalDraftUploadStarter.upload`
- [x] 443666a1627f6cea5f18cf2f712e79db2bfe0185 Added error handling which I forgot. I'm not sure how to add a test for this though.